contracts/bridge/RateLimiter.sol

```diff
@@ -0,0 +1,264 @@
+        if (resetBaseMin > 0) {
+            require(resetBaseMin <= currentMin, "resetBaseMin <= currentMin");
+            allowance.lastResetMin =
+                currentMin -
+                ((currentMin - resetBaseMin) % resetTimeMin);
+        } else if (allowance.lastResetMin == 0) {
+            allowance.lastResetMin = currentMin;
+        }
+        allowance.resetTimeMin = resetTimeMin;
+        allowance.amount = allowanceAmount;
+        updateAllowance(token, allowance);
+        emit SetAllowance(token, allowanceAmount, resetTimeMin);
+    }
+
+    function updateAllowance(address token, Allowance memory allowance)
+        private
+    {
+        allowances[token] = allowance;
+    }
+
+    function getAllowance(address token)
+        private
+        view
+        returns (Allowance memory allowance)
+    {
+        allowance = allowances[token];
+        // solium-disable-next-line security/no-block-members
+        uint32 currentMin = uint32(block.timestamp / 60);
+        // Check if we should reset the time. We do this on load to minimize storage read/ writes
+        if (
+            allowance.resetTimeMin > 0 &&
+            allowance.lastResetMin <= currentMin - allowance.resetTimeMin
+        ) {
+            allowance.spent = 0;
+            // Resets happen in regular intervals and `lastResetMin` should be aligned to that
+            allowance.lastResetMin =
+                currentMin -
+                ((currentMin - allowance.lastResetMin) %
+                    allowance.resetTimeMin);
+        }
+        return allowance;
+    }
+
+    /**
+     * @notice Checks the allowance for a given token. If the new amount exceeds the allowance, it is not updated and false is returned
+     * otherwise true is returned and the transaction can proceed
+     * @param amount to transfer
+     **/
+    function checkAndUpdateAllowance(address token, uint256 amount)
+        external
+        nonReentrant
+        onlyRole(BRIDGE_ROLE)
+        returns (bool)
+    {
+        Allowance memory allowance = getAllowance(token);
+
+        // Update state
+        // @dev reverts if amount > (2^96 - 1)
+        uint96 newSpent = allowance.spent + uint96(amount);
+
+        // Check overflow
+        require(
+            newSpent > allowance.spent,
+            "overflow detected: newSpent > allowance.spent"
+        );
+
+        // do not proceed. Store the transaction for later
+        if (newSpent >= allowance.amount) {
+            return false;
+        }
+
+        allowance.spent = newSpent;
+        updateAllowance(token, allowance);
+
+        return true;
+    }
+
+    function addToRetryQueue(bytes32 kappa, bytes memory toRetry)
+        external
+        onlyRole(BRIDGE_ROLE)
+    {
+        rateLimited.set(kappa, toRetry);
+    }
+
+    function retryByKappa(bytes32 kappa) external onlyRole(LIMITER_ROLE) {
+        (bytes memory toRetry, ) = rateLimited.get(kappa);
```

Comment:
Check if kappa already exists before calling bridge. I think assumption that **all kappas** in the queue were not yet processed is too strong to preserve.
```solidity
if (!IBridge(BRIDGE_ADDRESS).kappaExists(kappa)) {
    (bool success, bytes memory returnData) = BRIDGE_ADDRESS.call(toRetry);
    require(
        success,
        Strings.append("could not call bridge:", _getRevertMsg(returnData))
    );
}
rateLimited.remove(kappa);
```